### PR TITLE
docs: Introduces documentation for secret detection within repos

### LINKED ISSUE #778: Create guidance around secret detection within repos
Add guidance around incorporating secret detection into new or existing repos.

### PR DESCRIPTION
**Pull request by wbreza**
# Pull Request Template

## What are you trying to address

- Provides good practices for enabling a robust secret detection workflow as part of your teams software development lifecycle.
- Resolves #778 

## Description of new changes

- Adds docs around incorporating secret detection workflow into repos
- References existing repos for enabling secret detection using a combination of:
  - Pre-commit hooks
  - Github workflows / AzDO pipelines

## For all pull requests

- [x] Link to the issue you are solving (so it gets closed)
- [x] Label the pull request with the appropriate area(s)
- [x] Assign potential reviewers (you may also want to contact them on Teams to ensure timely reviews)
- [x] Assign the project - Engineering Playbook Backlog
- [x] Assign the pull request to yourself

### CODE REVIEW

File: docs/security/secret-detection.md [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+Following software assurance security best practices can easily deter secrets from entering a codebase.
+Ensure your application data is not compromised by introducing a simple and easy way to detect secrets during its development, and gate for secrets in your CI/CD pipelines.
+Today we introduce to you an open-source secrets detection framework [Yelp Detect-Secrets](https://github.com/Yelp/detect-secrets), brought to you by Yelp.
+This tool utilizes a Shannon Entropy model to detect secrets in codebases and prevents new secrets from being entered.
+This guide will demonstrate two easy to use tools using the Detect-Secrets framework brought to you by the CSE Engineering Fundamentals and Security group to keep secrets from entering from pre-commit and pre-production environment.
+
+## Detect Secrets Pre-Commit Hook
+
+The Detect-Secrets pre-commit hook is a tool that keeps secrets from being entering the codebase at one of the earliest stages of development.
+This tool will run a detect-secrets scan right before a Git commit occurs and if any findings occur, then mitigation can occur before and history of the secret is generated in Git history.
```
Comment (TessFerrandez) [inline docs/security/secret-detection.md:12]:
```suggestion
This tool will run a detect-secrets scan right before a Git commit occurs and if any findings occur, then mitigation can occur before any history of the secret is generated in Git history.
```

File: docs/security/secret-detection.md [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+Using the Detect Secrets Pre-Commit approach developers can insure they don't accidentally commit secrets at the earliest point in the development cycle but at a project level it requires all developers to adopt this approach to provide protection.
```
Comment (TessFerrandez) [inline docs/security/secret-detection.md:13]:
```suggestion
Using the Detect Secrets Pre-Commit approach developers can ensure they don't accidentally commit secrets at the earliest point in the development cycle but at a project level it requires all developers to adopt this approach to provide protection.
```

File: docs/security/secret-detection.md [hunk 3]
```diff
@@ -0,0 +1,111 @@
+# Secret Detection
+
+Following software assurance security best practices can easily deter secrets from entering a codebase.
+Ensure your application data is not compromised by introducing a simple and easy way to detect secrets during its development, and gate for secrets in your CI/CD pipelines.
+Today we introduce to you an open-source secrets detection framework [Yelp Detect-Secrets](https://github.com/Yelp/detect-secrets), brought to you by Yelp.
+This tool utilizes a Shannon Entropy model to detect secrets in codebases and prevents new secrets from being entered.
+This guide will demonstrate two easy to use tools using the Detect-Secrets framework brought to you by the CSE Engineering Fundamentals and Security group to keep secrets from entering from pre-commit and pre-production environment.
```
Comment (TessFerrandez) [inline docs/security/secret-detection.md:7]:
```suggestion
It is important to ensure that secrets don't make it in to the code base and compromise the security of our application data. The best way of doing so is by introducing a simple and easy way to detect secrets during its development, and gate for secrets in your CI/CD pipelines.
This section describes how we can use the open-source secrets detection framework [Yelp Detect-Secrets](https://github.com/Yelp/detect-secrets), to accomplish this together with two tools from the Detect-Secrets framework created by the CSE Engineering Fundamentals and Security group.
The Yelp Detect-Secrets tool utilizes a Shannon Entropy model to detect secrets in codebases and prevents new secrets from being entered.
```

Comment (TessFerrandez) [reply docs/security/secret-detection.md:7]: Some language suggestions... take them or leave them. The original felt a bit selly to me :) but that is just my opinion, so its ok with me if you want to keep the original.

File: docs/security/secret-detection.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+## Detect Secrets Pre-Commit Hook
+
+The Detect-Secrets pre-commit hook is a tool that keeps secrets from being entering the codebase at one of the earliest stages of development.
+This tool will run a detect-secrets scan right before a Git commit occurs and if any findings occur, then mitigation can occur before and history of the secret is generated in Git history.
+Using the Detect Secrets Pre-Commit approach developers can insure they don't accidentally commit secrets at the earliest point in the development cycle but at a project level it requires all developers to adopt this approach to provide protection.
+
+### Configuration for Environment
+
+Below contain two ways of setting up your environment depending on your use case.
```
Comment (TessFerrandez) [inline docs/security/secret-detection.md:17]:
```suggestion
Below are two ways of setting up your environment depending on your use case.
```

File: docs/security/secret-detection.md [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+- [Pre-Commit Hook within Existing Repo](https://github.com/wbreza/pre-commit-hooks/blob/main/detect-secrets/README.md)
+  - Simple install script bootstraps existing repos with pre-commit hooks and helper scripts to help manage the secret detection lifecycle
+- [Pre-Commit Hook with New Repo Template](https://github.com/wbreza/baseline-security-seed/blob/main/SECURITY.md)
+  - Contains all of the above plus out of the box support for dev containers using the Github universal dev container image.
+  - Also includes a GitHub workflow to additionally perform secret detection within pull requests
+
+## Detect Secrets Azure DevOps Marketplace Extension
+
+This extension scans the code base for secrets by using sophisticated algorithms utilizing Shannon entropy and reports the findings into the Azure DevOps Pipeline model to enable quick feedback and response from development teams throughout the development life cycle.
```
Comment (TessFerrandez) [inline docs/security/secret-detection.md:27]:
```suggestion
This extension scans the code base for secrets using sophisticated algorithms such as Shannon entropy and reports the findings into the Azure DevOps Pipeline model to enable quick feedback and response from development teams throughout the development life cycle.
```

File: docs/security/secret-detection.md [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+As an extension deployed in the CI/CD pipeline Detect Secrets does not rely on developers to change their development environment.
+Depending on how your pipeline is configured replying on the CI/CD extension can lead to situations where secrets are leaked into the repository before being detected.
+Leaking secrets into source the repository does require cleanup which is why using both pre-commit and CI/CD protections are recommended.
+
+### Configuration
+
+Currently this extension is in private preview and to get access reach out to a member of the Security Solution Area and we will onboard you and your crew.
```
Comment (TessFerrandez) [inline docs/security/secret-detection.md:34]: Small reminder that this is a public repo that is used outside of CSE as well... so we should avoid using terms like the Security Solution Area etc. Also, prefer the more generic term Team over Crew - as Crew is very CSE specific

**Review by TessFerrandez — APPROVED** [review]
Great content... had some language suggestions - and also a reminder that while this content is targeted at CSE - we should avoid mention of internal structure (like solution areas), crews etc. as this is also meant to be a resource to leave behind with our partners, and also for use by the general public.

I have approved with suggestions

**Comment by brockneedscoffee** [comment]
@wbreza do you want me to move the content in dev-sec-ops to this section once this is merged?  A lot of this content already exists there and then some.

**Review by brockneedscoffee — APPROVED** [review]

**Review by TessFerrandez — APPROVED** [review]

**Review by shiranr — APPROVED** [review]

**Pull request merged**